controllers/healthz/health_reconciler.go

```diff
@@ -0,0 +1,172 @@
+// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package healthz
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/gardener/controller-manager-library/pkg/logger"
+	druidv1alpha1 "github.com/gardener/etcd-druid/api/v1alpha1"
+	"github.com/gardener/etcd-druid/pkg/client/kubernetes"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	// reconcilerName is the name of the controller
+	reconcilerName = "healthz_reconciler"
+	healthyLabel   = "healthy"
+)
+
+// Reconciler is health recocniler object for external service created on etcd
+type Reconciler struct {
+	clietnset kubernetes.Interface
+	client    client.Client
+}
+
+// NewHealthReconciler return the health reconciler object for external service created on etcd
+func NewHealthReconciler(cli kubernetes.Interface) reconcile.Reconciler {
+	return &Reconciler{
+		clietnset: cli,
+		client:    cli.Client(),
+	}
+}
+
+// Reconcile reconciles the etcd health.
+func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.TODO()
+
+	etcd := &druidv1alpha1.Etcd{}
+	if err := r.client.Get(ctx, req.NamespacedName, etcd); err != nil {
+		if errors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	logger.WithLogger(ctx, "reconciler", reconcilerName)
+	logger.Infof("Reconciling etcd for health: %s/%s", etcd.GetNamespace(), etcd.GetName())
+	if etcd.Status.ServiceName == nil {
+		logger.Infof("missing external service. Skipping reconciliation.")
+		return ctrl.Result{
+			RequeueAfter: time.Second * 5,
+		}, nil
+	}
+
+	// TODO: Discuss getBackupHealthOverExec vs getBackupHealthOverHTTP
+	healthy, err := r.getBackupHealthOverHTTP(ctx, etcd)
```

Comment:
Is it a good idea to do the health check inline during reconciliation? This would health status being up-to-date subject to the vagaries of concurrency. Any delay in setting an endpoint as unhealthy could be costly and a delay in setting it to be healthy will mean prolonging the downtime. In the dependency-watchdog probe, I erred on the side of dedicated probe goroutines per shoot. WDYT?

Reply:
> Is it a good idea to do the health check inline during reconciliation? This could lead health status being up-to-date subject to the vagaries of concurrency. 

Maybe i didn't get your point completely.
If you mean to do health check in goroutine for all etcd nodes concurrently, I agree that will reduce the time delay in  setting health status. 

> Any delay in setting an endpoint as unhealthy could be costly and a delay in setting it to be healthy will mean prolonging the downtime.

Yeah, agreed. This is gonna be tough job; updating health status in minimum time, so that endpoint will be removed or added asap.

Reply:
Yes. I meant that if we do the check inline in reconciliation, then the amount of concurrency is limited by the number of worker routines which probably will always be smaller than the number of `etcds` in a seed cluster. There is a similar problem that dependency-watchdog probe command also faces. I erred on the side of launching a dedicated goroutine per target shoot cluster that is probed with a mechanism to restart those probes if any configuration of the shoots changed. You can see the relevant part [here](https://github.com/gardener/dependency-watchdog/blob/master/pkg/scaler/scaler.go#L237-L244).